designs/2018-async-commands/README.md

```diff
@@ -0,0 +1,81 @@
+- Start Date: 2018-12-2
+- RFC PR: (leave this empty, to be filled in later)
+- Authors: David Aghassi
+
+# Async/Parallizing CLIEngine
+
+## Summary
+
+`CLIEngine` currently doesn't support any async functions. As such, large codebases pay a toll since it requires users to wait for each file ( time O(n) ) to be processed by ESLint. There's a large gain to be made if ESLint can offer parallelized tasks so that users can more quickly statically analyze their code.
+
+## Motivation
+
+On code bases in larger companies (in my case, Intuit) ESLint can take a very long time relative to other build tasks, even with caching enabled. The faster users can iterate and get feedback for their code, the faster they can develop/debug/ship. It also has impacts on CI since the same toll is paid.
+
+## Detailed Design
+
+Based on discussion in the original [PR](https://github.com/eslint/eslint/issues/10606#issue-341171187)there are some main points.
+
+1. The features that are intended to be async/parallel need to be behind a flag on the CLI, and opt in when using the API (maybe through an async command or optoin when using CLIEngine).
+2. There are two options here: `promise` based and `thread` based. Some users suggested looking at [esprint](https://github.com/pinterest/esprint) for inspiration. Some of the concerns were that `promises` would not be faster per say depending on the implementation. On the flipside, it has been noted that `threads`/workers can be expensive on platforms like Windows. Ideally, the implementation would use a balance of both. For example, `jest` has async methods that use threads under the hood.
+3. The experience should be opt in so as not to break existing users.
+4. There should be some sort of mechanism to handle eslint rules that expect `sync` reading. Rules may need to identify themselves as a "type". This way, if a rule is being loaded that requires `sync`, the code can handle it accordingly. Any rule that doesn't identify as a `sync` or `async` would be defaulted to `sync`.
```

Comment:
Do you have any examples that demonstrate why this is necessary? I'm not sure I follow.

Reply:
I was pulling this idea from this comment https://github.com/eslint/eslint/issues/10606#issuecomment-424864137

I honestly don't know the inner workings of a rule set, so I don't know that rules can necessarily be `sync` or `async` by nature.

Reply:
Gotcha. Currently, those visitor callbacks are synchronous (see https://github.com/eslint/eslint/issues/10606#issuecomment-424883533 and https://github.com/eslint/eslint/issues/10606#issuecomment-425471740).

Reply:
Agreed. There is no such thing as an async rule, and even if there was, that should be handled by something lower in the API stack. I think it's safe to remove this reference.